### ISSUE TITLE
Lock mechanism can stick in infinite loop

### ISSUE THREAD

**Issue by ainlolcat** [jplana/python-etcd#181]
Hello. I have code like :

```
client = etcd.Client(host='etcd', protocol='http', port=2379)   
lock = etcd.Lock(client, 'lock_key')  
print 'Try to acquire Lock'  
lock.acquire(blocking=blocking, lock_ttl=lock_ttl, timeout=timeout)  
```

And sometimes it just freezes on lock.acquire(...). Logs indicate what etcd.client tries to read value over and over: 

> ...
> 161 Try to acquire Lock
> 162 2016-06-09 14:19:53,213 - etcd.client - DEBUG - Issuing read for key /_locks/lock_key with args {'recursive': True}
> 163 2016-06-09 14:19:53,241 - urllib3.connectionpool - DEBUG - "GET /v2/keys/_locks/lock_key?recursive=true HTTP/1.1" 200 109
> 164 2016-06-09 14:19:53,242 - etcd.lock - DEBUG - Lock not found, writing it to /_locks/lock_key
> 165 2016-06-09 14:19:53,243 - etcd.client - DEBUG - Writing 7e932ac6c2ae428aae0bcac0aa73017b to key /_locks/lock_key ttl=1 dir=False append=True
> 166 2016-06-09 14:19:53,441 - urllib3.connectionpool - DEBUG - "POST /v2/keys/_locks/lock_key HTTP/1.1" 201 210
> 167 2016-06-09 14:19:53,468 - etcd.lock - DEBUG - Lock key /_locks/lock_key/35619 written, sequence is 35619
> 168 2016-06-09 14:19:53,471 - etcd.client - DEBUG - Issuing read for key /_locks/lock_key with args {'recursive': True}
> 169 2016-06-09 14:19:53,477 - urllib3.connectionpool - DEBUG - "GET /v2/keys/_locks/lock_key?recursive=true HTTP/1.1" 200 485
> 170 2016-06-09 14:19:53,478 - etcd.lock - DEBUG - Lock keys found: [u'/_locks/lock_key/35618', u'/_locks/lock_key/35619']
> 171 2016-06-09 14:19:53,560 - etcd.lock - DEBUG - Locker: /_locks/lock_key/35618, key to watch: /_locks/lock_key/35618
> 172 2016-06-09 14:19:53,560 - etcd.lock - DEBUG - Lock not acquired, now watching /_locks/lock_key/35618
> 173 2016-06-09 14:19:53,561 - etcd.client - DEBUG - About to wait on key /_locks/lock_key/35618, index None
> 174 2016-06-09 14:19:53,566 - etcd.client - DEBUG - Issuing read for key /_locks/lock_key/35618 with args {'recursive': None, 'timeout': 1, 'wait': True}
> 175 2016-06-09 14:19:53,569 - urllib3.connectionpool - DEBUG - "GET /v2/keys/_locks/lock_key/35618?wait=true HTTP/1.1" 200 None
> 176 2016-06-09 14:19:54,468 - etcd.lock - DEBUG - Detected variation for /_locks/lock_key/35618: expire
> 177 2016-06-09 14:19:54,469 - etcd.client - DEBUG - Issuing read for key /_locks/lock_key with args {'recursive': True}
> 178 2016-06-09 14:19:54,471 - urllib3.connectionpool - DEBUG - "GET /v2/keys/_locks/lock_key?recursive=true HTTP/1.1" 200 109
> 179 2016-06-09 14:19:54,472 - etcd.lock - DEBUG - Lock keys found: [u'/_locks/lock_key']
> 180 2016-06-09 14:19:54,472 - etcd.client - DEBUG - About to wait on key /_locks/lock_key/35618, index None
> 181 2016-06-09 14:19:54,472 - etcd.client - DEBUG - Issuing read for key /_locks/lock_key/35618 with args {'recursive': None, 'timeout': 1, 'wait': True}
> 182 2016-06-09 14:19:54,474 - urllib3.connectionpool - DEBUG - "GET /v2/keys/_locks/lock_key/35618?wait=true HTTP/1.1" 200 None
> 183 2016-06-09 14:19:55,476 - etcd.client - DEBUG - Watch timed out.
> 184 2016-06-09 14:19:55,477 - etcd.client - DEBUG - About to wait on key /_locks/lock_key/35618, index None
> 185 2016-06-09 14:19:55,477 - etcd.client - DEBUG - Issuing read for key /_locks/lock_key/35618 with args {'recursive': None, 'timeout': 1, 'wait': True}
> 186 2016-06-09 14:19:55,477 - urllib3.connectionpool - INFO - Resetting dropped connection: etcd
> 187 2016-06-09 14:19:55,625 - urllib3.connectionpool - DEBUG - "GET /v2/keys/_locks/lock_key/35618?wait=true HTTP/1.1" 200 None
> 188 2016-06-09 14:19:56,628 - etcd.client - DEBUG - Watch timed out.
> 189 2016-06-09 14:19:56,628 - etcd.client - DEBUG - About to wait on key /_locks/lock_key/35618, index None
> 190 2016-06-09 14:19:56,628 - etcd.client - DEBUG - Issuing read for key /_locks/lock_key/35618 with args {'recursive': None, 'timeout': 1, 'wait': True}
> 191 2016-06-09 14:19:56,629 - urllib3.connectionpool - INFO - Resetting dropped connection: etcd
> 192 2016-06-09 14:19:56,646 - urllib3.connectionpool - DEBUG - "GET /v2/keys/_locks/lock_key/35618?wait=true HTTP/1.1" 200 None
> 193 2016-06-09 14:19:57,648 - etcd.client - DEBUG - Watch timed out.
> 194 2016-06-09 14:19:57,683 - etcd.client - DEBUG - About to wait on key /_locks/lock_key/35618, index None
> 195 2016-06-09 14:19:57,683 - etcd.client - DEBUG - Issuing read for key /_locks/lock_key/35618 with args {'recursive': None, 'timeout': 1, 'wait': True}
> ...

I beleave it can happen because of https://github.com/jplana/python-etcd/blob/master/src/etcd/lock.py#L127 

> ```
>          while True:
>             try:
>                 r = self.client.watch(watch_key, timeout=t)
>                 _log.debug("Detected variation for %s: %s", r.key, r.action)
>                 return self._acquired(blocking=True, timeout=timeout)
>             except etcd.EtcdKeyNotFound:
>                 _log.debug("Key %s not present anymore, moving on", watch_key)
>                 return self._acquired(blocking=True, timeout=timeout)
>             except etcd.EtcdException:
>                 # TODO: log something...
>                 pass
> ```

At first I think it is unlikely because I miss second log entry from _get_locker() method but it can be what _get_locker will fail with exception,  "except etcd.EtcdException" part will catch exception and start watch method again and again and then code will fail into infinite loop. May be it is normal behavior but I specify timeout and want result or error after timeout. There are two problems: 
- Someone ignores exception. You can log anything You like but please log something. It will be much easier to understand and debug
- Someone ignores timeout. I wrap execution and add timeout for now.

**Comment by obi1kenobi** [comment]
The lock code is awful and I think it should be removed. Here is a trivial situation that will reliably cause deadlock:

From one python shell (call it A), run:
```
client = etcd.Client()
lock = etcd.Lock(client, lock_name='foo')
lock.acquire(blocking=True, lock_ttl=10)
time.sleep(10)
lock.release()
```

Immediately after starting the above, from another python shell (call it B) run:
```
client = etcd.Client()
lock = etcd.Lock(client, lock_name='foo')
lock.acquire(blocking=True, lock_ttl=5)
```

Note that the second lock call has a TTL of 5, while the first has a TTL of 10. The following set of events happen, in order:
- A successfully acquires the lock with a TTL of 10 seconds
- B doesn't manage to acquire a lock, but gets the next key in the sequence, with a TTL of 5 seconds
- B starts to watch A's key for changes
- B's key expires, but B doesn't see that because it's only watching A's key
- A's key expires, A exits
- from this point all execution is on B: B sees A's key expire and [reads the list of keys](https://github.com/jplana/python-etcd/blob/master/src/etcd/lock.py#L159); since there are no keys under the lock, the lock itself is a leaf and is returned: `[u'/_locks/foo']`
- the `l.index(self.lock_key)` [call](https://github.com/jplana/python-etcd/blob/master/src/etcd/lock.py#L165) doesn't find the lock key, causing a `ValueError` which is then reraised as an `EtcdLockExpired` error
- this error is caught in the `except EtcdException: pass` [block](https://github.com/jplana/python-etcd/blob/master/src/etcd/lock.py#L126) and simply causes a retry of the watch on A's key
- since A's key doesn't exist, the watch never returns because there will never be any changes on the key -- *deadlock*.

**Comment by ainlolcat** [comment]
We hit problems with key removal actually. Looks like this:
1. 1st thread create lock record 1
2. 1st thread does some stuff
3. 2nd thread looks at locks list and find record 1. Tries to create record 2
4. 1st thread complete task and removes record 1
5. 2nd thread starts to wait record 1 in infinite loop because record does not exits.

Actually we fix problem with little sed script:

`    sed -i -e 's/except ValueError:/except ValueError as val_err:/g' /usr/lib/python2.7/site-packages/python_etcd-0.4.3-py2.7.egg/etcd/lock.py && \
    sed -i -e 's/raise etcd.EtcdLockExpired(u"Lock not found")/_log.error(val_err)\n            raise etcd.EtcdLockExpired(u"Lock not found")/g'  /usr/lib/python2.7/site-packages/python_etcd-0.4.3-py2.7.egg/etcd/lock.py && \
    sed -i -e 's/return (l\[0\], l\[i-1\])/return (l\[0\], next(x for x in results if x.key == l\[i-1\]))/g' /usr/lib/python2.7/site-packages/python_etcd-0.4.3-py2.7.egg/etcd/lock.py && \
    sed -i -e 's/watch_key = nearest/watch_key = nearest.key/g' /usr/lib/python2.7/site-packages/python_etcd-0.4.3-py2.7.egg/etcd/lock.py && \
    sed -i -e 's/r = self.client.watch(watch_key, timeout=t)/r = self.client.watch(watch_key, timeout=t, index=nearest.modifiedIndex + 1)/g' /usr/lib/python2.7/site-packages/python_etcd-0.4.3-py2.7.egg/etcd/lock.py && \
    cat /usr/lib/python2.7/site-packages/python_etcd-0.4.3-py2.7.egg/etcd/lock.py && \
    rm -rf /usr/lib/python2.7/site-packages/python_etcd-0.4.3-py2.7.egg/etcd/lock.pyc`

So main change is what we watch record 1 from last point we see it so if cluster has changes after this point (deletion in our case) it will return result instead of timeout.

**Comment by ainlolcat** [comment]
I solution provided above is correct I can create pull request with it.

**Comment by obi1kenobi** [comment]
I believe your solution is correct for the problem you pointed out, but does not solve the problem I pointed out. Would you mind testing the scenario I constructed and seeing if it still deadlocks?

**Comment by fantasylzp** [comment]
could anyone fixed it  or has any other method ?

**Comment by ainlolcat** [comment]
Scenario suggested by @obi1kenobi leads to stuck. My solution does not cover it but 2 lines can solve it. To reproduce @obi1kenobi  scenario execute

``` python
import time
import etcd
import threading


def run_long_lock():
    client = etcd.Client()
    lock = etcd.Lock(client, lock_name='foo')

    print '[{}] Try to acquire long Lock'.format(time.time())
    lock.acquire(blocking=True, lock_ttl=10)
    print '[{}] Long Lock acquired'.format(time.time())

    time.sleep(10)

    print '[{}] Try to release long Lock'.format(time.time())
    lock.release()
    print '[{}] Long Lock released'.format(time.time())


def run_short_lock():
    client = etcd.Client()
    lock = etcd.Lock(client, lock_name='foo')
    print '[{}] Try to acquire short Lock'.format(time.time())
    lock.acquire(blocking=True, lock_ttl=5)
    print '[{}] Short Lock acquired'.format(time.time())
    lock.release()


long_thread = threading.Thread(target=run_long_lock)
long_thread.start()

# to make sure if long lock is ready
time.sleep(1)

short_thread = threading.Thread(target=run_short_lock)
short_thread.start()
short_thread.join()
```

**Comment by ainlolcat** [comment]
I have created pull request https://github.com/jplana/python-etcd/pull/206

**Comment by ainlolcat** [comment]
Forgot to post test case for error with lock race. It's simple but anyway. I start etcd docker container like this:

```
export HostIP="192.168.12.50"
docker run -d -p 4001:4001 -p 2380:2380 -p 2379:2379 \
  --name etcd \
  quay.io/coreos/etcd \
  /usr/local/bin/etcd \
  -advertise-client-urls http://${HostIP}:2379,http://${HostIP}:4001 \
  -listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001
```

And then start python script:

```python
import threading
import time
import etcd

client = etcd.Client(host='localhost', protocol='http', port=2379)


def run_lock_getter(name="Thread1"):
    while True:
        lock = etcd.Lock(client, 'lock_key')
        print '[{}:{}] Try to acquire Lock'.format(name, time.time())
        lock.acquire()
        print '[{}:{}] Try to release Lock'.format(name, time.time())
        lock.release()


first_thread = threading.Thread(target=run_lock_getter, args=("Thread1", ))
first_thread.start()

second_thread = threading.Thread(target=run_lock_getter, args=("Thread2", ))
second_thread.start()
second_thread.join()
```

It will acquire/release lock in 2 loops and will stuck after several seconds. Issue remains until I applied fix above.

**Comment by jplana** [comment]
After a few tests, I see the problems that both @obi1kenobi and @ainlolcat describe here, and the seem to be fixed by @ainlolcat PR #206. Probably locking is not completely fixed by this, but at least this deadlock is fixed.

**Comment by jplana** [comment]
BTW, thank you both for the great description of the problem!
